subsys/bluetooth/services/bas/bas_bls.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2024 Demant A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/bluetooth/services/bas.h>
+#include <zephyr/bluetooth/gatt.h>
+#include "bas_internal.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(bas, CONFIG_BT_BAS_LOG_LEVEL);
+
+/* The battery level status of a battery. */
+static struct bt_bas_bls battery_level_status;
+
+/*
+ * Bitfield structure: Power State
+ *
+ * - Bits 0: Battery Present
+ * - Bits 1–2: Wired External Power Source Connected
+ * - Bits 3–4: Wireless External Power Source Connected
+ * - Bits 5–6: Battery Charge State
+ * - Bits 7–8: Battery Charge Level
+ * - Bits 9–11: Charging Type
+ * - Bits 12–14: Charging Fault Reason
+ * - Bit 15: RFU
+ *
+ * For detailed specification, refer to:
+ * https://bitbucket.org/bluetooth-SIG/public/src/main/gss/
+ * org.bluetooth.characteristic.battery_level_status.yaml
+ */
+#define BATTERY_SHIFT              0
+#define WIRED_POWER_SHIFT          1
+#define WIRELESS_POWER_SHIFT       3
+#define BATTERY_CHARGE_STATE_SHIFT 5
+#define BATTERY_CHARGE_LEVEL_SHIFT 7
+#define BATTERY_CHARGE_TYPE_SHIFT  9
+#define CHARGING_FAULT_SHIFT       12
+
+/*
+ * Bitfield structure: Additional Status
+ *
+ * - Bits 0–1: Service Required
+ * - Bit 2: Battery Fault
+ * - Bits 3–7: Reserved
+ */
+#define SERVICE_REQUIRED_SHIFT 0
+#define BATTERY_FAULT_SHIFT    2
+
+void bt_bas_bls_init(void)
+{
+	LOG_INF("Initialise BAS Battery Level Status Module");
+
+	battery_level_status.flags = 0;
+	battery_level_status.power_state = 0;
+
+#if defined(CONFIG_BT_BAS_BLS_IDENTIFIER_PRESENT)
+	/* Set identifier flag */
+	battery_level_status.flags |= BT_BAS_BLS_IDENTIFIER_PRESENT;
+	battery_level_status.identifier = 0;
+#endif
+
+#if defined(CONFIG_BT_BAS_BLS_BATTERY_LEVEL_PRESENT)
+	/* Set battery level flag */
+	battery_level_status.flags |= BT_BAS_BLS_BATTERY_LEVEL_PRESENT;
+	battery_level_status.battery_level = bt_bas_get_battery_level();
+#endif
+
+#if defined(CONFIG_BT_BAS_BLS_ADDITIONAL_STATUS_PRESENT)
+	/* Set additional status flag */
+	battery_level_status.flags |= BT_BAS_BLS_ADDITIONAL_STATUS_PRESENT;
+	battery_level_status.additional_status = 0;
+#endif
+}
+
+ssize_t read_blvl_status(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			 uint16_t len, uint16_t offset)
+{
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &battery_level_status,
+				 sizeof(struct bt_bas_bls));
```

Comment:
Does BAS mandate that the data is little endian? If so, can we just provide the pointer like this for big endian devices?